File: src/78.子集.rs
/*
 * @lc app=leetcode.cn id=78 lang=rust
 *
 * [78] 子集
 */
struct Solution;
// @lc code=start
impl Solution {
    pub fn subsets(nums: Vec<i32>) -> Vec<Vec<i32>> {
        let n = nums.len();
        let mut ret = vec![];
        for i in 0..(1 << n) {
            let mut sub = vec![];
            for j in 0..n {
                if (i >> j) & 1 == 1 {
                    sub.push(nums[j]);
                }
            }
            ret.push(sub);
        }
        ret
    }
}
// @lc code=end
mod utils;
#[test]
fn test_solution() {
    macro_rules! test {
        ($args:tt, $ans:tt) => {
            assert_eq!(
                Solution::subsets(vec!$args),
                vec2d!$ans
            )
        };
    }
    test!(
        [1, 2, 3],
        [[], [1], [2], [1, 2], [3], [1, 3], [2, 3], [1, 2, 3]]
    );
}

File: src/78.子集.py
#
# @lc app=leetcode.cn id=78 lang=python3
#
# [78] 子集
#
from typing import List
from utils import *
# @lc code=start


class Solution:
    def subsets(self, nums: List[int]) -> List[List[int]]:
        n = len(nums)
        ans = []
        for k in range(2 ** n):
            subset = []
            for i in range(n):
                if (k >> i) & 1 == 1:
                    subset.append(nums[i])
            ans.append(subset)
        return ans

# @lc code=end


if __name__ == '__main__':
    def test(input, expected):
        calc = Solution().subsets(input)
        if set(map(tuple, calc)) != set(map(tuple, expected)):
            print(f'case failed: `{input}`')
            print(f'  calc     = `{calc}`')
            print(f'  expected = `{expected}`')
            exit(1)
    test([1, 2, 3], [
        [3],
        [1],
        [2],
        [1, 2, 3],
        [1, 3],
        [2, 3],
        [1, 2],
        []
    ])
